modules/dnn/src/graph_simplifier.cpp

```diff
@@ -202,6 +202,7 @@ void simplifySubgraphs(const Ptr<ImportGraphWrapper>& net,
             {
                 patterns[j]->replace(net, matchedNodesIds, targetNodesIds);
                 numNodes -= matchedNodesIds.size() - 1;  // #matchedNodes removed and one added.
+                i -= matchedNodesIds.size() - 1;
```

Comment:
I don't think that it works as expected.

Am I right that after fusion of Gather+Cost or Mul+Const layers you want to run matching for Upsample layer? the problem is that it's patter start a bit earlier (includes at least Shape node) that means we cannot shift enough back to match Upsample after Gather.

So maybe we need to make the first loop over patterns but internal one by nodes? This way we can guaratnee that each following patter in our list will start from the beginning of graph. But we need to benchmark the performance of graphs fusion logic - how much it will grow up.

Reply:
Just tested two approaches for graphs matching on Mask R-CNN network (1401 nodes, 17 TensorFlow patterns to match)

1) current
```cpp
    for (int i = 0; i < numNodes; ++i)
    {
        for (int j = 0; j < patterns.size(); ++j)
        {
            if (patterns[j]->match(net, i, matchedNodesIds, targetNodesIds))
            {
                patterns[j]->replace(net, matchedNodesIds, targetNodesIds);
                numNodes -= matchedNodesIds.size() - 1;  // #matchedNodes removed and one added.
                break;
            }
        }
    }
```
21.3925ms

2) Modified:
```cpp
    for (int j = 0; j < patterns.size(); ++j)
    {
        for (int i = 0; i < numNodes; ++i)
        {
            if (patterns[j]->match(net, i, matchedNodesIds, targetNodesIds))
            {
                patterns[j]->replace(net, matchedNodesIds, targetNodesIds);
                numNodes -= matchedNodesIds.size() - 1;  // #matchedNodes removed and one added.
            }
        }
    }
```
22.6679ms

So I think you can freely try to use it for your experiment without any significant efficiency gaps.

Reply:
Below graph is not a part of any test, I have exported this graph just to show why I was adding ` i -= matchedNodesIds.size() - 1;`
![Screenshot from 2020-03-02 08-17-33](https://user-images.githubusercontent.com/26292819/75641521-8b2ec200-5c5e-11ea-8a75-d490015cdd8c.png)
Net nodes before graph simplification

```
Conv Constant Shape Gather Cast Constant Mul Cast Floor Constant Shape Gather Cast Constant Mul Cast Floor Unsqueeze Unsqueeze Concat Constant Cast Shape Slice Cast Div Concat Upsample 
Conv Constant Shape Gather Cast Constant Mul Cast Floor Constant Shape Gather Cast Constant Mul Cast Floor Unsqueeze Unsqueeze Concat Constant Cast Shape Slice Cast Div Concat Upsample
```
Net nodes before graph simplification

if `i -= matchedNodesIds.size() - 1;` is **not added** (not desired)
```
Conv Constant Constant Constant Constant Constant Upsample 
Conv Constant Shape Gather Cast Constant Mul Cast Floor Constant Shape Gather Cast Constant Mul Cast Floor Unsqueeze Unsqueeze Concat Constant Cast Shape Slice Cast Div Concat Upsample
```
so here, after fusion at first `upsample` node, nodes ids changed but` i` was still the same, hence it missed `Cast` after `Gather` and `Mul` nodes because ids of these nodes became less than `i` which in turn results in graph mismatch at second `upsample `node so it didn't fuse. 

if `i -= matchedNodesIds.size() - 1;` is **added** (desired)
```
Conv Constant Constant Constant Constant Constant Upsample 
Conv Constant Constant Constant Constant Constant Upsample
```

In my view, if graph would be like 
```
                                  | Unfused_upsample |
                                           |
                                  | Unfused_softmax |
```
then it would miss unfused_softmax fusion after unfused_upsample fusion if i is not changed. 
Please correct me if I misinterpreted something.

Reply:
Thanks, This modified version is working perfectly fine
```cpp
for (int j = 0; j < patterns.size(); ++j)
    {
        for (int i = 0; i < numNodes; ++i)
        {
            if (patterns[j]->match(net, i, matchedNodesIds, targetNodesIds))
            {
                patterns[j]->replace(net, matchedNodesIds, targetNodesIds);
                numNodes -= matchedNodesIds.size() - 1;  // #matchedNodes removed and one added.
            }
        }
    }
```

Reply:
So will you consider this change in this PR? I believe that this way we can avoid `i -= matchedNodesIds.size() - 1;`

Reply:
yeah, I will add this change here.

Reply:
> So if you don't mind, let's move a test from PyTorch 1.2. to your PR (see opencv/opencv_extra#716).

@dkurt, here you wanted me to Push a PR to Opencv_extra adding your suggested change to that new PR?

Reply:
Yes, they should be bonded, thanks!